src/Umbraco.Infrastructure/Migrations/Upgrade/V_13_0_0/AddGuidsToUsers.cs

```diff
@@ -0,0 +1,262 @@
+﻿using NPoco;
+using Umbraco.Cms.Core;
+using Umbraco.Cms.Infrastructure.Persistence.DatabaseAnnotations;
+using Umbraco.Cms.Infrastructure.Persistence.DatabaseModelDefinitions;
+using Umbraco.Cms.Infrastructure.Persistence.Dtos;
+using Umbraco.Cms.Infrastructure.Scoping;
+using Umbraco.Extensions;
+
+namespace Umbraco.Cms.Infrastructure.Migrations.Upgrade.V_13_0_0;
+
+/// <summary>
+/// This is an unscoped migration to support migrating sqlite, since it doesn't support adding columns.
+/// See <see cref="AddGuidsToUserGroups"/> for more information.
+/// </summary>
+public class AddGuidsToUsers : UnscopedMigrationBase
+{
+    private const string NewColumnName = "key";
+    private readonly IScopeProvider _scopeProvider;
+
+    public AddGuidsToUsers(IMigrationContext context, IScopeProvider scopeProvider)
+        : base(context)
+    {
+        _scopeProvider = scopeProvider;
+    }
+
+    protected override void Migrate()
+    {
+        using IScope scope = _scopeProvider.CreateScope();
+        using IDisposable notificationSuppression = scope.Notifications.Suppress();
+        ScopeDatabase(scope);
+
+        if (DatabaseType != DatabaseType.SQLite)
+        {
+            MigrateSqlServer();
+            scope.Complete();
+            return;
+        }
+
+        MigrateSqlite();
+        scope.Complete();
+    }
+
+    private void MigrateSqlServer()
+    {
+        var columns = SqlSyntax.GetColumnsInSchema(Context.Database).ToList();
+        AddColumnIfNotExists<UserDto>(columns, NewColumnName);
+        List<UserDto>? dtos = Database.Fetch<UserDto>();
+        if (dtos is null)
+        {
+            return;
+        }
+
+        MigrateExternalLogins(dtos);
+        MigrateTwoFactorLogins(dtos);
+    }
+
+    private void MigrateSqlite()
+    {
+        if (ColumnExists(Constants.DatabaseSchema.Tables.User, NewColumnName))
+        {
+            return;
+        }
+
+        // See AddGuidsToUserGroups for more information on why we're doing this.
+        Database.Execute("COMMIT;");
+        Database.Execute("PRAGMA foreign_keys=off;");
+        Database.Execute("BEGIN TRANSACTION;");
```

Comment:
Are we ever committing this transaction and reenable the foreing keys?

Reply:
This is a weird one, but it's due to how SQLite and our DB implementation work. 

We commit this transaction when we complete the scope since we call `ScopeDatabase`, this will use the DB from the scope for the database operation, meaning when our scope is completed the transaction is too.


And we don't explicitly reenable the foreign keys, since this happens automatically the next time a connection is opened, SQLite defaults to foreign keys being on every time a connection is established. This is also why we have to do this in a weird way by using SQL directly instead of `Database.CompleteTransaction();` since that'll also close the connection, undoing the work we just did.

Reply:
Sounds like good info to have in a code comment

Reply:
Very good point, I've added a small paragraph outline why this is done 👍